### PR TITLE
Add Detekt distribution support

### LINKED ISSUE #14: User-provided Detekt CLI Jar
To allow user-specific updates and or forks of Detekt.

This is easy to support as is, but API might change with Persistent Worker support #7 due to potential changes in classloading/etc, thus it's on hold

### PR DESCRIPTION
**Pull request by arturdryomov**
Closes #14.

### CODE REVIEW

File: detekt/distribution.bzl [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+load("@rules_jvm_external//:defs.bzl", "maven_install")
+
+def detekt_distribution(version = "1.2.2"):
+    maven_install(
+        artifacts = [
+            "io.gitlab.arturbosch.detekt:detekt-cli:{v}".format(v = version),
+        ],
+        repositories = [
+            "https://jcenter.bintray.com/",
+            "https://repo1.maven.org/maven2",
```
Comment (artem-zinnatullin) [inline detekt/distribution.bzl]: Reorder pls, Maven Central is more stable than JCenter

File: detekt/distribution.bzl [hunk 2]
```diff
@@ -0,0 +1,20 @@
+"""
+A macro for changing Detekt distribution.
+"""
+
+load("@rules_jvm_external//:defs.bzl", "maven_install")
+
+def detekt_distribution(version = "1.2.2"):
+    maven_install(
```
Comment (artem-zinnatullin) [inline detekt/distribution.bzl]: I'm not exactly sure what's the story with overriding maven_install inflated from other WORKSPACE, do you want to try to to override detekt to a custom jar to test if that's actually possible?

**Comment by artem-zinnatullin** [comment]
Don't forget to update the "Upcoming features:" in README when you verify that this way actually makes it possible, thx

**Comment by arturdryomov** [comment]
@artem-zinnatullin, all right, so I’ve checked like you’ve asked in a new workspace.

```python
http_archive(
    name = "rules_detekt",
    urls = ["https://github.com/buildfoundation/bazel_rules_detekt/archive/ad/distribution.zip"],
    strip_prefix = "bazel_rules_detekt-ad-distribution",
    sha256 = "5f7ade79bcfa3f9a0a220fea99c6c82579ae984e1b81a4f12bf9ba3b9df6563b",
)

load("@rules_detekt//detekt:distribution.bzl", "detekt_distribution")

detekt_distribution()
```
```python
load("@rules_detekt//detekt:defs.bzl", "detekt")

detekt(
    name = "sample",
    srcs = glob(["src/main/kotlin/**/*.kt"]),
)
```
```console
$ bazelisk build //:sample

Regular results here.
```
```diff
- detekt_distribution()
+ detekt_distribution(version = "1.2.1")
```
```console
$ bazelisk build //:sample --worker_verbose

INFO: Analyzed target //:sample (3 packages loaded, 183 targets configured).
INFO: Found 1 target...
INFO: Destroying Detekt worker (id 5)
INFO: Created new non-sandboxed Detekt worker (id 6), logging to /private/var/tmp/_bazel_ming/e1928b54ea884af1b0a78e98fd2799db/bazel-workers/worker-6-Detekt.log
Target //:sample up-to-date:
  bazel-bin/sample_detekt_report.txt
INFO: Elapsed time: 13.453s, Critical Path: 4.09s
INFO: 9 processes: 7 darwin-sandbox, 2 worker.
INFO: Build completed successfully, 13 total actions
```
```diff
- detekt_distribution(version = 1.2.1)
+ detekt_distribution(version = "1.2.2")
```
```console
$ bazelisk build //:sample --worker_verbose

INFO: Analyzed target //:sample (3 packages loaded, 183 targets configured).
INFO: Found 1 target...
INFO: Destroying Detekt worker (id 6)
INFO: Created new non-sandboxed Detekt worker (id 7), logging to /private/var/tmp/_bazel_ming/e1928b54ea884af1b0a78e98fd2799db/bazel-workers/worker-7-Detekt.log
Target //:sample up-to-date:
  bazel-bin/sample_detekt_report.txt
INFO: Elapsed time: 11.952s, Critical Path: 2.10s
INFO: 2 processes: 2 worker.
INFO: Build completed successfully, 6 total actions
```

So it seems to work.

FYI — the macros idea is not new. See `rules_kotlin` [documentation](https://github.com/bazelbuild/rules_kotlin#custom-kotlinc-distribution-and-version) and [implementation](https://github.com/bazelbuild/rules_kotlin/blob/master/kotlin/internal/repositories/repositories.bzl).

One thing though. I had to declare `rules_kotlin`, `rules_jvm_external` and `rules_proto` to make the sample work. And seems like it is [an expected behavior](https://docs.bazel.build/versions/master/external.html#transitive-dependencies). Do we need to mention it in the readme or Bazel people are familiar enough with this “feature”?

**Comment by Kernald** [comment]
A common pattern for transitive dependencies is to provide a macro to call, responsible for checking if the dependencies are already declared, and if not, declare them. See an example [here](https://github.com/bazelbuild/rules_go/blob/master/go/private/repositories.bzl). This way allows to have a simple drop-in setup, while still allowing the consumers to override any repository if needed.

**Comment by arturdryomov** [comment]
@Kernald, nice, thanks! I’ll explore it in a separate PR.

**Comment by arturdryomov** [comment]
@artem-zinnatullin, requesting your review again since I’m not sure I’ve checked what you actually asked for or it was something different.

**Comment by artem-zinnatullin** [comment]
I think this is good enough for now (def better than what it was), there might be an issue if someone wants to point Detekt to a custom target type (`http_file`, `java_library`, etc) instead of a maven dep because we unconditionally do `maven_install`, but until someone complaints I wouldn't bother, it's much more important that we decouple hardcoded Detekt version from rule releases

@Kernald thanks for commenting :)

**Pull request merged**